droidlet/tools/hitl/nsp_retrain_infra.py

```diff
@@ -0,0 +1,370 @@
+        except OSError:
+            logging.info(f"Likely error: sweep_runner.py not found where expected")
+            raise
+        except ValueError:
+            logging.info(f"Likely error: Popen called with invalid arguments")
+            raise
+
+        # Wait for child process to finish and log outputs/errors
+        now = datetime.now()
+        while (sweep.poll() != 0):
+            if ((datetime.now() - now).seconds > NSP_RETRAIN_TIMEOUT):
+                sweep.kill()
+                raise TimeoutError(f"NSP Retrain child process timed out after {NSP_RETRAIN_TIMEOUT} seconds")
+                break
+        outs, errs = sweep.stdout.read(), sweep.stderr.read()
+        logging.info(f"Sweep script outputs: \n{outs}")
+        logging.info(f"Sweep script errors: \n{errs}")
+
+        # Find the model output directory -- ASSUMES UNIQUE BATCH_ID
+        for dir in os.listdir(opts.output_dir):
+            if dir.startswith(batch_id):
+                output_models_dir = os.path.join(opts.output_dir, (dir + "/"))
+        try:
+            model_out = os.path.join(output_models_dir, "model_out/")
+        except:
+            logging.info(f"model output directory not found, check batch ID")
+            raise
+
+        # Retrieve cluster job IDs
+        job_ids = []
+        for line in outs.splitlines():
+            if 'Submitted' in line:
+                job_ids.append(line[-8:])
+
+        # Wait for slurm jobs to complete and determine the best model
+        logging.info(f"Watching slurm for job numbers: {job_ids}")
+        os.chdir(model_out)
+        for i in range(NUM_MODEL_OUTPUT_POLLS):
+            if ((datetime.now() - now).seconds > NSP_RETRAIN_TIMEOUT):
+                raise TimeoutError(f"NSP Retrain child process timed out after {NSP_RETRAIN_TIMEOUT} seconds")
+            if self.slurm_jobs_finished(job_ids):
+                accs = self.get_accs()
+                self.copy_best_model(accs)
+                break
+            time.sleep(MODEL_OUTPUT_POLL_TIME)
+
+        # Save the best model in S3 bucket and close out
+        upload_key = batch_id + "/best_model/best_model.pth" 
+        s3.upload_file("best_model.pth", 'droidlet-hitl', upload_key)
+
+        logging.info(f"NSP Retraining Job finished")
+        self.set_finished(True)
+        return
+
+
+class NSPNewDataListener(JobListener):
+    def __init__(self, batch_id, opts):
+        super(NSPNewDataListener, self).__init__()
+        self.batch_id = batch_id
+        self.new_data_found = False
+        self.opts = opts
+
+    def run(self, runner):
+        logging.info(f"NSP New Data Listener running")
+
+        while not self.check_is_finished():
+            time.sleep(LISTENER_SLEEP_TIME)
+            try:
+                prefix = str(self.batch_id) + '/'
+                new_data_key = s3.list_objects_v2(Bucket='droidlet-hitl', Prefix=prefix, Delimiter='/')['Contents'][1]['Key']
```

Comment:
Found this in trial run: don't assume there will only be one file named `meta.txt` under `bucket_name/batch_id` since it's somewhere we keep all run-specific files in. In fact, before meta.txt is uploaded to S3, there're a lot of other files already uploaded such as interaction logs, annotated commands, etc.
Let's just directly search for `bucket/batch_id/meta.txt` instead of checking file existence.